### ISSUE TITLE
Better support for separate data paths

### ISSUE THREAD

**Issue by tabokie** [pingcap/tiup#854]
## Feature Request

**Is your feature request related to a problem? Please describe:**

When using separate path for raftdb, tiup doesn't create the path if missing, also doesn't delete the path when cluster is destroyed.

Reproduce:
```
server_configs:
  tikv:
    raftstore.raftdb-path: "/data4/tabokie/tmp/raft"
```
deploy with error (tikv can't find the path):
```
Error: init config failed: 172.16.5.32:20187: executor.ssh.execute_failed: Failed to execute command over SSH for 'pingcap@172.16.5.32:22' {ssh_stderr: invalid configuration: entity not found
, ssh_stdout: , ssh_command: export LANG=C; PATH=$PATH:/usr/bin:/usr/sbin /data4/tabokie/tmp/tikv-20187/bin/tikv-server --config-check --config=/data4/tabokie/tmp/tikv-20187/conf/tikv.toml --pd=""}, cause: Process exited with status 1: check config failed
```

**Describe the feature you'd like:**

When config contains separate paths, take them into consideration on cluster setup/destroy.

**Describe alternatives you've considered:**

**Teachability, Documentation, Adoption, Migration Strategy:**

**Comment by 9547** [comment]
Currently, only those path will be created automatically:

- `deploy_dir`
- `data_dir`
- `log_dir`

We need to sort out the possible paths first, then create it if the yaml file specified, possible path as blow:

TiKV:

- `raftstore.raftdb-path`
- `rocksdb.wal-dir`
- `rocksdb.info-log-dir`
- `raftdb.info-log-dir`


TiDB:
...

**Comment by AstroProfundis** [comment]
There could be many of them, and it would be hard for tiup to follow all their changes in the future.

I think the best approach would be the *-server processes handle the directory creation, and only report error when the directory is not able to create or not able to write. This has one more issue: tiup is not aware of those directory created by processes, and may not properly delete them on scale-in.

If we let tiup handle the life cycle of directories, I'd prefer adding new fields to `*Spec` structs, instead of parsing `configs` map, this could give us more control of how the final config files are generated as well as the procedure of handling different directories.

**Comment by lucklove** [comment]
@tabokie Thanks for your issue, it's a good one.
But I should remind that this is a big challenge for TiUP: TiUP is responsible to cleanup the directories which created by it, so if TiUP creates directories specified in the config field of some servers, it should cleanup it in the scale-in or destroy/cleanup stage.
However, the user may change the config at any time (eg. change raftdb.info-log-dir from /tmp/a to /tmp/b and then to /tmp/c, we should cleanup them all), so there should be a trace mechanism to follow it or just forbid the user to change it.
What's worse is that the directory specified may be an existed one and already have data in it, TiUP doesn't know if it's created by TiUP or the user, if try to cleanup it, the data in it will be lost.

**Comment by tabokie** [comment]
I agree with @AstroProfundis to use tiup defined fields, like the way we did in `tidb-ansible`. We should still parse the config though to check possible conflict. (btw do we have this logic for data-dir?). And fail any attempt to change them through edit-config.

And for the issues mentioned by @lucklove, changing data dir (if supported at all) most definitely requires service downtime, simple cut-and-paste is enough without keeping track of history paths. Also, I think as long as user specifies the directory explicitly to tiup, we can assume exclusive ownership. Of course a heads-up would be fine.

**Comment by lucklove** [comment]
I think we should make all components (TiKV, PD, TiFlash, Drainer .etc) consistent:

- The component itself should create directory defined in it's config file
- And we should only defined these directories as a sub directory of `DeployDir` or `DataDir` if we expect TiUP to cleanup it

So in this case, I think the `TiKV` component should try to create the directory, just like drainer: https://github.com/pingcap/tiup/issues/957

**Comment by 9547** [comment]
I agree with @lucklove. The component itself should create the used directory(`DeployDir`, `DataDir` and `LogDir`).
